src/internal_rules.jl

```diff
@@ -558,3 +558,418 @@ function EnzymeRules.reverse(
+    elseif isa(fact, Const) && isa(B, Const)
+        nothing
+    else
+        error("Error in forward ldiv! Enzyme rule $(typeof(fact)) $(typeof(B)).")
+    end
+    if RT <: Const
+        return retval
+    elseif (RT <: DuplicatedNoNeed)
+        return retdval
+    else
+        return Duplicated(retval, retdval)
+    end
+end
+
+function EnzymeRules.forward(
+        func::Const{typeof(ldiv!)},
+        RT::Type{<:Union{Const, BatchDuplicatedNoNeed, BatchDuplicated}},
+        fact::Annotation{C}, B::Union{Const, BatchDuplicated{T,N}};
+        kwargs...
+) where {C <: Cholesky, T <: AbstractArray{<:AbstractFloat}, N}
+    retval = B.val
+    retval = ldiv!(fact.val, retval)
+    for b in 1:N
+        if isa(fact, BatchDuplicated) && isa(B, BatchDuplicated)
+            retdval = zeros(length(retval))
+            mul!(retdval, fact.dval[b].U, retval)
+            mul!(retdval, fact.dval[b].L, retdval)
+            retdval .= B.dval[b] .- retdval
+            ldiv!(fact.val, retdval)
+            B.val     .= retval
+            B.dval[b] .= retdval
+        elseif isa(fact, BatchDuplicated) && isa(B, Const)
+            retdval = zeros(length(retval))
+            mul!(retdval, A, B.val)
+            mul!(retdval, -1, retdval)
+            ldiv!(fact.val, retdval)
+            B.val     .= retval
+            B.dval[b] .= retdval
+        elseif isa(fact, Const) && isa(B, BatchDuplicated)
+            retdval = copy(B.dval)
+            ldiv!(fact.val, retdval)
+            B.val     .= retval
+            B.dval[b] .= retdval
+        elseif isa(fact, Const) && isa(B, Const)
+            nothing
+        else
+            error("Error in forward ldiv! Enzyme rule $(typeof(fact)) $(typeof(B)).")
+        end
+        retdval
+    end
+    if RT <: Const
+        return retval
+    elseif (RT <: DuplicatedNoNeed) || (RT <: BatchDuplicatedNoNeed)
+        return retdval
+    else
+        return BatchDuplicated(retval, ntuple(i -> bretdval[i], Val(N)))
+    end
+end
+
+function EnzymeRules.augmented_primal(
+    config,
+    func::Const{typeof(cholesky)},
+    RT::Type{<:Union{Const, Duplicated, BatchDuplicatedNoNeed, BatchDuplicated}},
+    A::Union{Const, Duplicated, BatchDuplicated};
+    kwargs...
+)
+    fact = if EnzymeRules.needs_primal(config)
+        cholesky(A.val; kwargs...)
+    else
+        nothing
+    end
+    # dfact would be a dense matrix, prepare buffer
+    dfact = if EnzymeRules.width(config) == 1
+        Cholesky(Matrix(fact), 'L', 0)
+    else
+        ntuple(Val(EnzymeRules.width(config))) do i
+            Base.@_inline_meta
+            Cholesky(Matrix(fact), 'L', 0)
+        end
+    end
+    return EnzymeRules.AugmentedReturn(fact, dfact, (dfact,))
+end
+
+function EnzymeRules.reverse(
+    config,
+    ::Const{typeof(cholesky)},
+    dret,
+    cache,
+    A;
+    kwargs...
+)
+    (_dfact,) = cache
+    for b in 1:EnzymeRules.width(config)
+        dA = if EnzymeRules.width(config) == 1
```

Comment:
see elsewhere, but we've found it simpler to do something like

```julia
dAs = if EnzymeRules.width(config) == 1
  (A.dval,)
else
  A.dval
end
```

and then have the for loop be 
```julia
for (dA, dfact) in zip(dAs, dfacts)
  ...
end
```